File: 2021/java/856.score-of-parentheses.java
class Solution {
    // solution 1: recursion
    // base case: () -- return 1
    // case 1: (A)(B) -> (A) + (B) 
    // case 2: ((A)) -> 2 * (A) 
    public int scoreOfParentheses(String S) {
        return helper(S, 0, S.length() - 1);
    }
    private int helper(String s, int l, int r) {
        if (r - l == 1) return 1; // () base case
        int count = 0;
        for (int i = l; i < r; i++) {
            if (s.charAt(i) == '(') count++;
            else if (s.charAt(i) == ')') count--;
            if (count == 0) { // balanced, (A)(B) -> (A) + (B)
                return helper(s, l, i) + helper(s, i + 1, r);
            }
        }
        // case (((A))) -> 2 * ((A))
        return 2 * helper(s, l + 1, r - 1); 
    }
    
    // solution 2: layer by layer travese
    // (()(())) -> (()) + ((())) = 2^1 + 2^2 = 6
    // () -> 2^0 = 1
    // (()()) -> (()) + (()) -> 2^1 + 2^1 = 4
    public int scoreOfParentheses(String S) {
        int res = 0;
        int lvl = 0;
        int len = S.length();
        for (int i = 0; i < len; i++) {
            char ch = S.charAt(i);
            if (ch == '(') lvl++;
            else lvl--;
            if (ch == ')' && S.charAt(i - 1) == '(') {
                res += 1 << lvl;
            }
            
        }
        return res;
    }
    
}

File: 2021/python/856.score-of-parentheses.py
"""
856. Score of Parentheses

Given a balanced parentheses string S, compute the score of the string based on the following rule:

() has score 1
AB has score A + B, where A and B are balanced parentheses strings.
(A) has score 2 * A, where A is a balanced parentheses string.
 

Example 1:

Input: "()"
Output: 1
Example 2:

Input: "(())"
Output: 2
Example 3:

Input: "()()"
Output: 2
Example 4:

Input: "(()(()))"
Output: 6
 

Note:

S is a balanced parentheses string, containing only ( and ).
2 <= S.length <= 50

"""
def scoreOfParentheses(self, S: str) -> int:
    if not S:
        return 0
    res, lvl = 0, 0
    for i in range(len(S)):
        lvl += 1 if S[i] == '(' else -1
        if S[i] == ')' and S[i - 1] == '(':
            res += 1 << lvl
    return res
